test/test_jit_autocast.py

```diff
@@ -819,6 +819,27 @@ def test_nhwc_autocast_jit_trace_model(model, x):
                 continue
             test_nhwc_autocast_jit_trace_model(self.models[i], self.inputs[i])
 
+    def test_cat_promote(self):
+        class TestModel(torch.nn.Module):
+            def __init__(self):
+                super(TestModel, self).__init__()
+            def forward(self, a, b):
+                return torch.cat([a, b], 0)
+
+        test_model = TestModel().eval()
+        with torch.cpu.amp.autocast(cache_enabled=False, dtype=torch.bfloat16), torch.no_grad():
+            a = torch.rand(24, 128, 128)
+            b = torch.rand(24, 128, 128, dtype=torch.bfloat16)
+            c = test_model(a, b)
+            traced = torch.jit.trace(test_model, (a, b))
+        traced = torch.jit.freeze(traced)
+        for _ in range(3):
+            c2 = traced(a, b)
+        self.assertTrue(c.dtype, torch.float32)
+        self.assertTrue(c2.dtype, torch.float32)
+        traced_graph = traced.graph_for(a, b)
+        self.assertTrue(any(n.kind() == "aten::to" for n in traced_graph.nodes()))
```

Comment:
The fix works but the UT is failing because `traced_graph.nodes()` returns the `prim::If` node, as NNC is enabled here & creates a fusion group:

```
test_cat_promote (__main__.TestJitTraceAutocast) ... graph(%self : __torch__.___torch_mangle_1.TestModel,
      %a : Tensor,
      %b : Tensor):
  %3 : Float(24, 128, 128, strides=[16384, 128, 1], requires_grad=0, device=cpu), %4 : BFloat16(24, 128, 128, strides=[16384, 128, 1], requires_grad=0, device=cpu), %5 : bool = prim::TypeCheck[types=[Float(24, 128, 128, strides=[16384, 128, 1], requires_grad=0, device=cpu), BFloat16(24, 128, 128, strides=[16384, 128, 1], requires_grad=0, device=cpu)]](%a, %b)
  %6 : Tensor = prim::If(%5)
    block0():
      %7 : Float(48, 128, 128, strides=[16384, 128, 1], requires_grad=0, device=cpu) = prim::TensorExprGroup_0(%3, %4)
      -> (%7)
    block1():
      %8 : Function = prim::Constant[name="fallback_function", fallback=1]()
      %9 : (Tensor) = prim::CallFunction(%8, %a, %b)
      %10 : Tensor = prim::TupleUnpack(%9)
      -> (%10)
  return (%6)
with prim::TensorExprGroup_0 = graph(%a : Float(24, 128, 128, strides=[16384, 128, 1], requires_grad=0, device=cpu),
      %b : BFloat16(24, 128, 128, strides=[16384, 128, 1], requires_grad=0, device=cpu)):
  %2 : int = prim::Constant[value=0]()
  %3 : NoneType = prim::Constant()
  %4 : bool = prim::Constant[value=0]()
  %5 : int = prim::Constant[value=6]()
  %6 : Float(24, 128, 128, strides=[16384, 128, 1], requires_grad=0, device=cpu) = aten::to(%b, %5, %4, %4, %3) # test/test_jit_autocast.py:827:0
  %7 : Tensor[] = prim::ListConstruct(%a, %6)
  %8 : Float(48, 128, 128, strides=[16384, 128, 1], requires_grad=0, device=cpu) = aten::cat(%7, %2) # test/test_jit_autocast.py:827:0
  return (%8)
```

Reply:
Maybe we can disable NNC for this UT.

Reply:
Fixed. Thanks!